File: backend/internal/models/speech.go
package models

// speech.go — 全平台语音输入的浏览器通信协议模型
//
// 本文件只定义“浏览器与TE-DNA后端”之间的稳定协议，不包含：
//   1. 火山引擎私有二进制协议；
//   2. 数据库配置读取；
//   3. WebSocket连接管理；
//   4. 具体业务输入框状态。
//
// 语音输入的职责边界：
//   - 老师点击麦克风后，浏览器向后端发送一条start控制消息；
//   - 后续以WebSocket二进制消息发送16kHz、单声道、16bit PCM音频；
//   - 老师停止录音时发送stop控制消息；
//   - 后端把豆包返回的识别结果转换为下方统一事件；
//   - 前端只把识别文字写进当前输入框，绝不自动发送给AI。
//
// 隐私边界：
//   - 本协议不要求浏览器上传真实姓名、设备序列号或其它身份信息；
//   - 音频只在当前WebSocket连接中流转，不作为业务数据落库；
//   - 服务端日志不得记录完整音频和完整识别正文。

// ==================== 浏览器控制动作 ====================

const (
	// SpeechActionStart 表示浏览器准备开始发送音频。
	SpeechActionStart = "start"

	// SpeechActionStop 表示浏览器已经停止采集音频。
	// 后端收到后向豆包发送最后一个音频包，并等待最终识别结果。
	SpeechActionStop = "stop"

	// SpeechActionCancel 表示用户主动取消本次语音输入。
	// 取消后不要求等待最终识别结果。
	SpeechActionCancel = "cancel"
)

// ==================== 服务端事件类型 ====================

const (
	// SpeechEventReady 表示后端和豆包上游连接均已建立，
	// 浏览器可以开始发送PCM音频包。
	SpeechEventReady = "ready"

	// SpeechEventPartial 表示临时识别结果。
	// 后续结果可能覆盖或修正本次文字。
	SpeechEventPartial = "partial"

	// SpeechEventFinal 表示本次录音的最终识别结果。
	SpeechEventFinal = "final"

	// SpeechEventError 表示当前语音识别连接发生错误。
	SpeechEventError = "error"

	// SpeechEventClosed 表示连接已正常关闭。
	SpeechEventClosed = "closed"
)

// ==================== 浏览器请求模型 ====================

// SpeechStartRequest 浏览器开始录音时发送的JSON控制消息。
//
// 当前正式协议固定要求：
//   - 16000Hz；
//   - 16bit；
//   - 单声道；
//   - PCM little-endian。
//
// 字段仍由浏览器显式提交，便于服务端发现客户端实现错误，
// 但服务端不得因为客户端提交其它值而自动猜测或静默转换。
type SpeechStartRequest struct {
	Action        string `json:"action"`
	SampleRate    int    `json:"sample_rate"`
	BitsPerSample int    `json:"bits_per_sample"`
	Channels      int    `json:"channels"`
}

// SpeechControlRequest stop和cancel使用的轻量控制消息。
type SpeechControlRequest struct {
	Action string `json:"action"`
}

// ==================== 服务端响应模型 ====================

// SpeechUtterance 单个语音分句。
//
// Definite=true表示豆包已经通过VAD与二遍识别确认该分句，
// 后续通常不会再修改该分句内容。
type SpeechUtterance struct {
	Text      string `json:"text"`
	StartTime int    `json:"start_time"`
	EndTime   int    `json:"end_time"`
	Definite  bool   `json:"definite"`
}

// SpeechRecognitionEvent 后端发给浏览器的统一语音事件。
//
// RequestID由TE-DNA生成，用于串联一次识别请求；
// LogID由火山引擎返回，只用于服务端和控制台排错；
// 浏览器不得把二者当作授权凭证。
type SpeechRecognitionEvent struct {
	Event      string            `json:"event"`
	Text       string            `json:"text,omitempty"`
	Final      bool              `json:"final,omitempty"`
	Utterances []SpeechUtterance `json:"utterances,omitempty"`
	DurationMS int               `json:"duration_ms,omitempty"`
	RequestID  string            `json:"request_id,omitempty"`
	LogID      string            `json:"log_id,omitempty"`
	Code       string            `json:"code,omitempty"`
	Message    string            `json:"message,omitempty"`
}

File: frontend/src/api/speech.ts
/**
 * speech.ts — TE-DNA统一语音输入WebSocket客户端
 *
 * 浏览器通信链：
 *   麦克风
 *     → 本站 /api/v1/speech/stream
 *     → TE-DNA Go后端
 *     → 豆包流式语音识别2.0
 *
 * 安全边界：
 * 1. 火山APP ID和Access Token永远不进入浏览器；
 * 2. WebSocket不能设置Authorization请求头，因此沿用现有SSE的query token；
 * 3. 音频只发送二进制帧，不写入sessionStorage；
 * 4. 识别文字只交给业务输入框，不自动调用AI。
 *
 * 生命周期约束：
 * 1. onClosed最多通知一次；
 * 2. 服务端error/final/closed均视为终态事件；
 * 3. 调用close()后不再把本地主动关闭误报为网络中断；
 * 4. 浏览器WebSocket握手失败时，只暴露稳定用户文案。
 */

export type SpeechEventType =
  | 'ready'
  | 'partial'
  | 'final'
  | 'error'
  | 'closed'

export interface SpeechUtterance {
  text: string
  start_time: number
  end_time: number
  definite: boolean
}

export interface SpeechRecognitionEvent {
  event: SpeechEventType
  text?: string
  final?: boolean
  utterances?: SpeechUtterance[]
  duration_ms?: number
  request_id?: string
  log_id?: string
  code?: string
  message?: string
}

export interface SpeechStreamHandlers {
  onReady?: (
    event: SpeechRecognitionEvent,
  ) => void
  onPartial?: (
    event: SpeechRecognitionEvent,
  ) => void
  onFinal?: (
    event: SpeechRecognitionEvent,
  ) => void
  onError?: (
    event: SpeechRecognitionEvent,
  ) => void
  onClosed?: (
    event?: SpeechRecognitionEvent,
  ) => void
  onUnexpectedClose?: (
    message: string,
  ) => void
}

export interface SpeechStreamConnection {
  isReady: () => boolean
  sendAudio: (
    pcm: ArrayBuffer,
  ) => void
  stop: () => void
  cancel: () => void
  close: () => void
}

function buildSpeechWebSocketURL(
  token: string,
): string {
  const protocol =
    window.location.protocol === 'https:'
      ? 'wss:'
      : 'ws:'

  const query =
    new URLSearchParams({
      token,
    })

  return (
    `${protocol}//${window.location.host}` +
    `/api/v1/speech/stream?${query.toString()}`
  )
}

function parseSpeechEvent(
  raw: unknown,
): SpeechRecognitionEvent | null {
  if (
    !raw ||
    typeof raw !== 'object'
  ) {
    return null
  }

  const candidate =
    raw as Partial<SpeechRecognitionEvent>

  const allowed:
    SpeechEventType[] = [
      'ready',
      'partial',
      'final',
      'error',
      'closed',
    ]

  if (
    typeof candidate.event !== 'string' ||
    !allowed.includes(
      candidate.event as SpeechEventType,
    )
  ) {
    return null
  }

  return candidate as SpeechRecognitionEvent
}

function sendControl(
  socket: WebSocket,
  action:
    | 'start'
    | 'stop'
    | 'cancel',
): void {
  if (
    socket.readyState !==
    WebSocket.OPEN
  ) {
    throw new Error(
      '语音连接尚未建立',
    )
  }

  const payload =
    action === 'start'
      ? {
          action,
          sample_rate: 16000,
          bits_per_sample: 16,
          channels: 1,
        }
      : {
          action,
        }

  socket.send(
    JSON.stringify(payload),
  )
}

/**
 * 建立本站语音连接。
 *
 * WebSocket打开后自动发送start控制消息；
 * 调用方必须等onReady后才能发送PCM。
 */
export function createSpeechStream(
  token: string,
  handlers: SpeechStreamHandlers,
): SpeechStreamConnection {
  const normalizedToken =
    token.trim()

  if (!normalizedToken) {
    throw new Error(
      '登录状态已失效，请重新登录',
    )
  }

  const socket =
    new WebSocket(
      buildSpeechWebSocketURL(
        normalizedToken,
      ),
    )

  let ready = false
  let closedByClient = false
  let terminalReceived = false
  let closedDelivered = false

  socket.binaryType = 'arraybuffer'

  const deliverClosed = (
    event?: SpeechRecognitionEvent,
  ) => {
    if (closedDelivered) return

    closedDelivered = true
    ready = false
    handlers.onClosed?.(event)
  }

  socket.onopen = () => {
    try {
      sendControl(
        socket,
        'start',
      )
    } catch {
      closedByClient = true

      try {
        socket.close()
      } catch {
        deliverClosed()
      }
    }
  }

  socket.onmessage = (
    message: MessageEvent,
  ) => {
    if (
      typeof message.data !==
      'string'
    ) {
      return
    }

    let raw: unknown

    try {
      raw = JSON.parse(
        message.data,
      )
    } catch {
      return
    }

    const event =
      parseSpeechEvent(raw)

    if (!event) return

    switch (event.event) {
      case 'ready':
        ready = true
        handlers.onReady?.(event)
        break

      case 'partial':
        handlers.onPartial?.(event)
        break

      case 'final':
        terminalReceived = true
        handlers.onFinal?.(event)
        break

      case 'error':
        terminalReceived = true
        handlers.onError?.(event)
        break

      case 'closed':
        terminalReceived = true
        deliverClosed(event)
        break
    }
  }

  socket.onerror = () => {
    /**
     * WebSocket API不会暴露握手HTTP响应正文。
     * 详细状态由close统一映射为稳定文案。
     */
  }

  socket.onclose = () => {
    ready = false

    if (
      !closedByClient &&
      !terminalReceived
    ) {
      handlers.onUnexpectedClose?.(
        '语音连接意外中断，请重新尝试',
      )
    }

    deliverClosed()
  }

  return {
    isReady: () =>
      ready &&
      socket.readyState ===
        WebSocket.OPEN,

    sendAudio: (
      pcm: ArrayBuffer,
    ) => {
      if (
        !ready ||
        socket.readyState !==
          WebSocket.OPEN
      ) {
        throw new Error(
          '语音识别尚未就绪',
        )
      }

      if (
        pcm.byteLength > 0
      ) {
        socket.send(pcm)
      }
    },

    stop: () => {
      sendControl(
        socket,
        'stop',
      )
    },

    cancel: () => {
      if (
        socket.readyState ===
        WebSocket.OPEN
      ) {
        try {
          sendControl(
            socket,
            'cancel',
          )
        } catch {
          // 连接正在关闭时无需重复报错。
        }
      }
    },

    close: () => {
      closedByClient = true
      ready = false

      if (
        socket.readyState ===
          WebSocket.OPEN ||
        socket.readyState ===
          WebSocket.CONNECTING
      ) {
        try {
          socket.close()
        } catch {
          deliverClosed()
        }
      } else {
        deliverClosed()
      }
    },
  }
}
